src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -396,13 +403,15 @@ class GroupReadsByUmi
     outHeader.setGroupOrder(GroupOrder.query)
     val out = SamWriter(output, outHeader)
 
-    val iterator = sorter.iterator.grouped(2).map(i => (i(0), i(1))).buffered // consume in pairs
+    val iterator = sorter.iterator.grouped(2).map { case Seq(x,y) => if (x.firstOfPair) (x, y) else (y, x) }.buffered // consume in pairs
     val tagFamilySizeCounter = new NumericCounter[Int]()
 
     while (iterator.hasNext) {
       // Take the next set of pairs by position and assign UMIs
       val pairs = takeNextGroup(iterator)
-      pairs.foreach { case(r1, r2) => assert(r1.name == r2.name, s"Reads out of order @ ${r1.id} + ${r2.id}") }
+      pairs.foreach { case(r1, r2) =>
+        assert(r1.name == r2.name && r1.firstOfPair && r2.secondOfPair, s"Reads out of order @ ${r1.id} + ${r2.id}")
```

Comment:
I actually can't.  It's there to catch any potential bug/regression, but between the filtering and sorting that's done and then the grouping on line 406 this assertion _should_ never fail.